gpflow/experimental/check_shapes/checker.py

```diff
@@ -49,17 +50,93 @@ def _as_parsed_tensor_spec(tensor_spec: TensorSpecLike, context: ErrorContext) -
         if isinstance(tensor_spec, tuple):
             for dim in tensor_spec:
                 if isinstance(dim, int):
-                    dimension_specs.append(ParsedDimensionSpec(dim, None, False))
+                    dimension_specs.append(ParsedDimensionSpec(dim, None, False, False))
                 else:
                     assert dim is None
-                    dimension_specs.append(ParsedDimensionSpec(None, None, False))
+                    dimension_specs.append(ParsedDimensionSpec(None, None, False, False))
         else:
             assert tensor_spec is None
-            dimension_specs.append(ParsedDimensionSpec(None, None, True))
+            dimension_specs.append(ParsedDimensionSpec(None, None, True, False))
         shape = ParsedShapeSpec(tuple(dimension_specs))
         return ParsedTensorSpec(shape, None)
 
 
+@dataclass
+class _ObservedDim:
+    """
+    Storage of observed size of a single dimension variable.
+    """
+
+    size: Optional[int] = None
+
+    def update(self, actual: Optional[int], broadcast: bool) -> bool:
```

Comment:
would `check_then_update` not be a better name if we use this method as part of an assert as [in this line](https://github.com/GPflow/GPflow/pull/1849/files#diff-b8b7cd3ef17cb6b9a4464a873ba9d26050f3cc6403361319b6da1b3b01ca67a8R271)?

Reply:
> would `check_then_update` not be a better name if we use this method as part of an assert as [in this line](https://github.com/GPflow/GPflow/pull/1849/files#diff-b8b7cd3ef17cb6b9a4464a873ba9d26050f3cc6403361319b6da1b3b01ca67a8R271)?

Fair call. I'll call it `check_and_update`, though.